File: a1go/core/memory.go
package core

import (
	"fmt"
	"sort"
)

type Allocator interface {
	Alloc(size Size, id Id)
	Dealloc(id Id)
	Compact()
}

type Memory struct {
	Size       Size
	FreeBlocks BlockList
	UsedBlocks BlockList
}

func NewMemory(size Size) Memory {
	freeBlocks := BlockList{}
	freeBlocks.Add(NewFree(Size(size), Address(0)))
	return Memory{size, freeBlocks, BlockList{}}
}

func (m *Memory) Alloc(size Size, id Id, fetch_block func(Size, BlockList) int) {
	block_index := fetch_block(size, m.FreeBlocks)
	block := m.FreeBlocks.Get(block_index)
	m.FreeBlocks.Remove(block_index)
	if block.Size > size {
		m.FreeBlocks.Add(Block{Free, Sub(block.Size, size), Add(block.Start_addr, size), block.End_addr})
	}
	m.UsedBlocks.Add(Block{id, size, block.Start_addr, Add(block.Start_addr, size) - 1})
}

func (m *Memory) Dealloc(id Id) {
	for i := 0; i < m.UsedBlocks.Len(); i++ {
		block := m.UsedBlocks.Get(i)
		if block.Id == id {
			m.UsedBlocks.Remove(i)
			newFree := block.AsFree()
			for j := 0; j < m.FreeBlocks.Len(); j++ {
				free_block := m.FreeBlocks.Get(j)
				if free_block.CanMerge(newFree) {
					m.FreeBlocks.Remove(j)
					block = block.Merge(free_block)
					break
				}
			}
			m.FreeBlocks.Add(block.AsFree())
			return
		}
	}
}

func (m *Memory) Compact() {
	for i := 0; i < m.FreeBlocks.Len(); i++ {
		block := m.FreeBlocks.Get(i)
		for j := 0; j < m.FreeBlocks.Len(); j++ {
			if i != j {
				other_block := m.FreeBlocks.Get(j)
				if block.End_addr == other_block.Start_addr {
					block.Size = Add(block.Size, other_block.Size)
					block.End_addr = other_block.End_addr
					m.FreeBlocks.Remove(j)
					i--
					break
				}
			}
		}
	}
}

func (m *Memory) Execute(cmd Cmd, fetch_block func(Size, BlockList) int) {
	switch cmd.Op {
	case Alloc:
		m.Alloc(Size(cmd.Size), Id(cmd.BlockID), fetch_block)
	case Dealloc:
		m.Dealloc(Id(cmd.BlockID))
	case Compact:
		m.Compact()
	}
}

func (m *Memory) String() string {
	var str string
	str += "Allocated Blocks:\n"
	for _, block := range m.UsedBlocks.blocks {
		str += block.String()
	}
	str += "Free Blocks:\n"
	for _, block := range m.FreeBlocks.blocks {
		str += block.String()
	}
	str += "Fragmentation:\n" + fmt.Sprintf("%.16f", m.Fragmentation()) + "\n"
	return str
}

func (m *Memory) All() []Block {
	used := m.UsedBlocks.blocks
	free := m.FreeBlocks.blocks

	sort.Slice(used, func(i, j int) bool {
		return used[i].Start_addr < used[j].Start_addr
	})
	sort.Slice(free, func(i, j int) bool {
		return free[i].Start_addr < free[j].Start_addr
	})

	all := append(used, free...)
	return all
}

func (m *Memory) AsByteArray() []byte {
	bytes := make([]byte, m.Size)
	all := m.All()

	for _, block := range all {
		for i := range Range(block.Start_addr, block.End_addr) {
			if block.Id == Free {
				bytes[i] = 0
			} else {
				bytes[i] = 1
			}
		}
	}
	return bytes
}

func (m *Memory) Fragmentation() float64 {
	free := m.FreeBlocks.blocks
	var largest_i Size
	for i, b := range free {
		if i == 0 || b.Size > largest_i {
			largest_i = Size(i)
		}
	}
	largest := free[largest_i]
	free_memory := m.CalcFreeMemory()
	return 1 - (float64(largest.Size) / float64(free_memory))
}

func (m *Memory) CalcFreeMemory() Size {
	free_memory := Size(0)
	for i := 0; i < m.FreeBlocks.Len(); i++ {
		block := m.FreeBlocks.Get(i)
		free_memory = Add(free_memory, block.Size)
	}
	return free_memory
}

func Range[I Address | int](start I, end I) []int {
	var rnge []int
	for i := start; i < end; i++ {
		rnge = append(rnge, int(i))
	}
	return rnge
}

File: a1rust/src/memory.rs
use std::io::Write;

use crate::{
    algos::{Algo, AlgoResult},
    block::{Address, Block, BlockVec, Id, Size},
    cmd::Cmd,
};

pub enum Result {
    Ok,
    AllocErr(Id, usize, usize),
    DeallocErr(Id, usize, usize),
}

impl std::fmt::Display for Result {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Result::Ok => write!(f, "OK"),
            Result::AllocErr(_, instr_nr, size) => write!(f, "A;{};{}", instr_nr, size),
            Result::DeallocErr(_, instr_nr, reason) => write!(f, "D;{};{}", instr_nr, reason),
        }
    }
}

pub struct Memory {
    pub size: usize,
    pub free_blocks: BlockVec,
    pub used_blocks: BlockVec,
    errors: Vec<Result>,
    instr_cnt: usize,
    out_cnt: usize,
}

impl Memory {
    pub fn new(size: usize) -> Memory {
        let mut free_blocks = BlockVec::new();
        free_blocks.add(Block::new_free(Size(size), Address(0)));
        Memory {
            size,
            free_blocks,
            used_blocks: BlockVec::new(),
            errors: vec![],
            instr_cnt: 0,
            out_cnt: 0,
        }
    }

    pub fn exec(&mut self, cmd: &Cmd, algo: Algo, path: &str) {
        self.incr();
        let res = match cmd {
            Cmd::Alloc(id, size) => self.alloc(Id(id.clone()), Size(size.clone()), algo),
            Cmd::Dealloc(id) => self.dealloc(Id(id.clone())),
            Cmd::Compact => self.compact(),
            Cmd::Output => {
                let out = self.output();
                let mut file =
                    std::fs::File::create(format!("{}.out{}", path, self.out_cnt)).unwrap();
                file.write_all(out.as_bytes()).unwrap();
                self.out_cnt += 1;
                Result::Ok
            }
        };

        match res {
            Result::Ok => (),
            _ => self.errors.push(res),
        }
    }

    pub fn free_memory(&self) -> usize {
        self.free_blocks.iter().fold(0, |acc, b| acc + b.size.0)
    }

    fn alloc(&mut self, id: Id, size: Size, func: fn(Size, &Vec<Block>) -> AlgoResult) -> Result {
        let index = func(size, &self.free_blocks.as_vec());
        match index {
            AlgoResult::Ok(i) => {
                let mut block = self.free_blocks.pop(i);
                let new_block = Block::new_used(id, size, block.start_addr);
                if new_block.size < block.size {
                    block.start_addr = block.start_addr + size;
                    block.size = block.size - size;
                    self.free_blocks.add(block);
                }
                self.used_blocks.add(new_block);
                Result::Ok
            }
            AlgoResult::None => Result::AllocErr(id, self.instr_cnt, self.free_memory()),
        }
    }

    fn dealloc(&mut self, id: Id) -> Result {
        for (i, b) in self.used_blocks.iter().enumerate() {
            if b.id.unwrap() == id {
                let mut block = self.used_blocks.pop(i);
                block.id = None;
                for (i, b) in self.free_blocks.iter().enumerate() {
                    if block.can_merge(b) {
                        let new_block = block.merge(b);
                        self.free_blocks.pop(i);
                        self.free_blocks.add(new_block);
                        return Result::Ok;
                    }
                }
                self.free_blocks.add(block);
                return Result::Ok;
            }
        }
        Result::DeallocErr(id, self.instr_cnt, self.did_try_allocating(id))
    }

    fn did_try_allocating(&self, id: Id) -> usize {
        let did_try: Option<&Result> = self.errors.iter().find(|e| match e {
            Result::AllocErr(i, _, _) => i == &id,
            _ => false,
        });

        match did_try {
            Some(Result::AllocErr(_, _, _)) => 1,
            _ => 0,
        }
    }

    fn compact(&mut self) -> Result {
        let mut nf_blocks = BlockVec::new();
        let mut nu_blocks = BlockVec::new();
        let f_mem = self.free_memory();

        let mut last_start_addr = Address(0);
        for block in self.used_blocks.iter() {
            let new_block =
                Block::new_used(block.id.unwrap(), block.size, last_start_addr + block.size);
            last_start_addr = new_block.end_addr + 1;
            nu_blocks.add(new_block);
        }
        nf_blocks.add(Block::new_free(Size(f_mem), Address(self.size - f_mem)));

        self.free_blocks = nf_blocks;
        self.used_blocks = nu_blocks;
        Result::Ok
    }

    pub fn output(&self) -> String {
        let mut out = format!("Size:\n{}\n", self.size);

        out.push_str("Allocated blocks:\n");
        for ub in self.used_blocks.iter() {
            out.push_str(&format!("{}\n", ub));
        }

        out.push_str("Free blocks:\n");
        for fb in self.free_blocks.iter() {
            out.push_str(&format!("{}\n", fb));
        }

        out.push_str(&format!("Fragmentation:\n{}\n", self.fragmentation()));

        out.push_str("Errors:\n");
        for err in self.errors.iter() {
            out.push_str(&format!("{}\n", err));
        }
        if self.errors.is_empty() {
            out.push_str("None\n");
        }
        out += "\n";

        out
    }

    fn fragmentation(&self) -> f64 {
        let largest_block = self.free_blocks.iter().max_by_key(|b| b.size).unwrap().size;
        let free_memory = self.free_memory();
        1f64 - largest_block.0 as f64 / free_memory as f64
    }

    fn incr(&mut self) {
        self.instr_cnt += 1;
    }
}
